### DISCUSSION TITLE
Typescript errors on stateToRoute and routeToState in stateMapping (InstantSearch Hooks routing)

### DISCUSSION THREAD

**Discussion by tom-radikal** [algolia/instantsearch#5519] [Q&A] [answered]
Hi,

I'm using Algolia in combination with Contentful and Next.js. One of the pages I'm developing is an archive page with a SearchBox and several RefinementLists. Since Algolia indexes the Contentful fields with its full name  (for example `fields.technology.en-US`), using the default `routing={true}` made the search query in the URL a bit too long.

Following the docs, I was using stateToRoute and routeToState to transform the necessary fields. Everything works as expected, but unfortunately only when I specifically add a ts-ignore comment above both these methods.

This is my InstantSearch component: 

```
<InstantSearch
  searchClient={searchClient}
  indexName={indexName}
  routing={{
    router: createInstantSearchRouterNext({
      singletonRouter,
    }),
    stateMapping: {
      stateToRoute(uiState) {
        const indexUiState = uiState[indexName]

        return {
          q: indexUiState.query,
          timing: indexUiState.refinementList?.timing,
          theme: indexUiState.refinementList?.['fields.theme.en-US'],
          type: indexUiState.refinementList?.['fields.type.en-US'],
          technology:
            indexUiState.refinementList?.['fields.technology.en-US'],
          sortBy: indexUiState.sortBy?.split(`${indexName}_`)[1],
          page: indexUiState.page,
        }
      },

      routeToState(routeState) {
        return {
          [indexName]: {
            query: routeState.q,
            refinementList: {
              timing: routeState.timing,
              'fields.theme.en-US': routeState.theme,
              'fields.type.en-US': routeState.type,
              'fields.technology.en-US': routeState.technology,
            },
            page: routeState.page,
            ...(routeState.sortBy
              ? { sortBy: `${indexName}_${routeState.sortBy}` }
              : {}),
          },
        }
      },
    },
  }}
>
```
And these are the Typescript errors I'm getting:
```
Type '(uiState: UiState) => { q: string | undefined; timing: string[] | undefined; theme: string[] | undefined; type: string[] | undefined; technology: string[] | undefined; sortBy: string | undefined; page: number | undefined; }' is not assignable to type '(uiState: UiState) => UiState'.
  Type '{ q: string | undefined; timing: string[] | undefined; theme: string[] | undefined; type: string[] | undefined; technology: string[] | undefined; sortBy: string | undefined; page: number | undefined; }' is not assignable to type 'UiState'.
    Property 'q' is incompatible with index signature.
      Type 'string | undefined' is not assignable to type 'Partial<{ query: string; } & { configure: PlainSearchParameters; } & { geoSearch: { boundingBox: string; }; } & { hierarchicalMenu: { [rootAttribute: string]: string[]; }; } & ... 13 more ... & { ...; }>'.
        Type 'undefined' is not assignable to type 'Partial<{ query: string; } & { configure: PlainSearchParameters; } & { geoSearch: { boundingBox: string; }; } & { hierarchicalMenu: { [rootAttribute: string]: string[]; }; } & ... 13 more ... & { ...; }>'.
```
```
Type '(routeState: UiState) => { [x: string]: { sortBy?: string | undefined; query: Partial<{ query: string; } & { configure: PlainSearchParameters; } & { geoSearch: { boundingBox: string; }; } & ... 14 more ... & { ...; }>; refinementList: { ...; }; page: Partial<...>; }; }' is not assignable to type '(routeState: UiState) => UiState'.
  Type '{ [x: string]: { sortBy?: string | undefined; query: Partial<{ query: string; } & { configure: algoliasearchHelper.PlainSearchParameters; } & { geoSearch: { boundingBox: string; }; } & ... 14 more ... & { ...; }>; refinementList: { ...; }; page: Partial<...>; }; }' is not assignable to type 'UiState'.
    'string' index signatures are incompatible.
      Type '{ sortBy?: string | undefined; query: Partial<{ query: string; } & { configure: algoliasearchHelper.PlainSearchParameters; } & { geoSearch: { boundingBox: string; }; } & { ...; } & ... 13 more ... & { ...; }>; refinementList: { ...; }; page: Partial<...>; }' is not assignable to type 'Partial<{ query: string; } & { configure: PlainSearchParameters; } & { geoSearch: { boundingBox: string; }; } & { hierarchicalMenu: { [rootAttribute: string]: string[]; }; } & ... 13 more ... & { ...; }>'.
        Types of property 'query' are incompatible.
          Type 'Partial<{ query: string; } & { configure: PlainSearchParameters; } & { geoSearch: { boundingBox: string; }; } & { hierarchicalMenu: { [rootAttribute: string]: string[]; }; } & ... 13 more ... & { ...; }>' is not assignable to type 'string'.
```

I must be doing something wrong obviously, but I can't seem to figure out what.

**Comment by Haroenv** [comment, marked as answer]
This probably could be documented more clearly indeed, but the RouteState defaults to UiState, so if you transform it, you should pass the generic to InstantSearch like this:


```tsx
import type { SearchClient, UiState } from 'instantsearch.js';
import React from 'react';
import { InstantSearch } from 'react-instantsearch-hooks';
import { createInstantSearchRouterNext } from 'react-instantsearch-hooks-router-nextjs';
import singletonRouter from 'next/router';

type RouteState = {
  q?: string;
  timing?: string[];
  theme?: string[];
  type?: string[];
  technology?: string[];
  sortBy?: string;
  page?: number;
};

export function App({
  indexName,
  searchClient,
}: {
  indexName: string;
  searchClient: SearchClient;
}) {
  return (
    <InstantSearch<UiState, RouteState>
      searchClient={searchClient}
      indexName={indexName}
      routing={{
        router: createInstantSearchRouterNext({
          singletonRouter,
        }),
        stateMapping: {
          stateToRoute(uiState) {
            const indexUiState = uiState[indexName];

            return {
              q: indexUiState.query,
              timing: indexUiState.refinementList?.timing,
              theme: indexUiState.refinementList?.['fields.theme.en-US'],
              type: indexUiState.refinementList?.['fields.type.en-US'],
              technology:
                indexUiState.refinementList?.['fields.technology.en-US'],
              sortBy: indexUiState.sortBy?.split(`${indexName}_`)[1],
              page: indexUiState.page,
            };
          },

          routeToState(routeState) {
            return {
              [indexName]: {
                query: routeState.q,
                refinementList: {
                  timing: routeState.timing,
                  'fields.theme.en-US': routeState.theme,
                  'fields.type.en-US': routeState.type,
                  'fields.technology.en-US': routeState.technology,
                },
                page: routeState.page,
                ...(routeState.sortBy
                  ? { sortBy: `${indexName}_${routeState.sortBy}` }
                  : {}),
              },
            };
          },
        },
      }}
    ></InstantSearch>
  );
}

```

**Reply by tom-radikal** [reply]
Did some more digging, and it seems the back button sometimes works, sometimes it doesn't. I tried setting strictMode to false, and removing all shared layout components, but to no avail.

**Reply by tom-radikal** [reply]
Could it maybe have something to do with this warning I'm getting in the console? I am using SearchBox, RefinementList and Pagination from `react-instantsearch-hooks-web`, so I'm not sure why I'm getting this warning.

<img width="1318" alt="image" src="https://user-images.githubusercontent.com/94045743/222230925-10bc816b-9da5-48b2-b98a-291cf563387b.png">

**Reply by tom-radikal** [reply]
I managed to get rid of the TS errors by making the properties in the refinementList required, and returning an empty Array in stateToRoute if it's undefined. This is my complete code for reference:

```
import styles from '../../styles/components/ProjectsArchive.module.scss'
import algoliasearch from 'algoliasearch/lite'
import { createInstantSearchRouterNext } from 'react-instantsearch-hooks-router-nextjs'
import { UiState } from 'instantsearch.js'
import {
  Configure,
  Hits,
  InstantSearch,
  Pagination,
  RefinementList,
  SearchBox,
  SortBy,
} from 'react-instantsearch-hooks-web'
import Wrapper from '../_atoms/Wrapper'
import Icon from '../_atoms/Icon'
import FunctionalButton from '../_atoms/FunctionalButton'
import { useState } from 'react'
import ProjectsArchiveHit from './ProjectsArchiveHit'
import singletonRouter from 'next/router'

const ProjectsArchive: React.FC = () => {
  const searchClient = algoliasearch(
    process.env.NEXT_PUBLIC_ALGOLIA_APPLICATION_ID as string,
    process.env.NEXT_PUBLIC_ALGOLIA_SEARCH_API_KEY as string
  )
  const indexName = `${process.env.NEXT_PUBLIC_ALGOLIA_ENVIRONMENT}_project`

  const [showFilters, setShowFilters] = useState(false)

  const toggleFilters = () => {
    setShowFilters(!showFilters)
  }

  type RouteState = {
    q?: string
    timing: string[]
    theme: string[]
    type: string[]
    technology: string[]
    sortBy?: string
    page?: number
  }

  return (
    <Wrapper>
      <InstantSearch<UiState, RouteState>
        searchClient={searchClient}
        indexName={indexName}
        routing={{
          router: createInstantSearchRouterNext({
            singletonRouter,
          }),
          stateMapping: {
            stateToRoute(uiState) {
              const indexUiState = uiState[indexName]

              return {
                q: indexUiState.query,
                timing: indexUiState.refinementList?.timing ?? [],
                theme:
                  indexUiState.refinementList?.['fields.theme.en-US'] ?? [],
                type: indexUiState.refinementList?.['fields.type.en-US'] ?? [],
                technology:
                  indexUiState.refinementList?.['fields.technology.en-US'] ??
                  [],
                sortBy: indexUiState.sortBy?.split(`${indexName}_`)[1],
                page: indexUiState.page,
              }
            },

            routeToState(routeState: RouteState) {
              return {
                [indexName]: {
                  query: routeState.q,
                  refinementList: {
                    timing: routeState.timing,
                    'fields.theme.en-US': routeState.theme,
                    'fields.type.en-US': routeState.type,
                    'fields.technology.en-US': routeState.technology,
                  },
                  page: routeState.page,
                  ...(routeState.sortBy
                    ? { sortBy: `${indexName}_${routeState.sortBy}` }
                    : {}),
                },
              }
            },
          },
        }}
      >
        <Configure
          attributesToSnippet={['excerpt:15']}
          snippetEllipsisText="…"
          hitsPerPage={12}
        />
        <div className={styles.container}>
          <div className={styles.side}>
            <div
              className={`${styles.filters} ${
                showFilters ? styles.showFilters : ''
              }`}
            >
              <Icon
                onClick={toggleFilters}
                className={styles.filterClose}
                icon="close"
              />
              <h3 className={styles.filterTitle}>Timing</h3>
              <RefinementList attribute="timing" sortBy={['name:asc']} />
              <h3 className={styles.filterTitle}>Theme</h3>
              <RefinementList
                attribute="fields.theme.en-US"
                sortBy={['name:asc']}
              />
              <h3 className={styles.filterTitle}>Funding</h3>
              <RefinementList
                attribute="fields.type.en-US"
                sortBy={['name:asc']}
              />
              <h3 className={styles.filterTitle}>Technology</h3>
              <RefinementList
                attribute="fields.technology.en-US"
                sortBy={['name:asc']}
              />
            </div>
          </div>
          <div className={styles.results}>
            <div className={styles.search}>
              <FunctionalButton
                text="Filter"
                type="secondary"
                scheme="functional"
                icon="filter"
                onClick={toggleFilters}
                className={styles.filterButton}
              />
              <SearchBox
                placeholder="Search"
                searchAsYouType={false}
                submitIconComponent={() => <Icon icon="search-false" />}
              />
              <SortBy
                items={[
                  { label: 'Most recent', value: `${indexName}` },
                  { label: 'Oldest first', value: `${indexName}_oldest` },
                ]}
              />
            </div>
            <div>
              <Hits hitComponent={ProjectsArchiveHit} />
            </div>
            <div>
              <Pagination />
            </div>
          </div>
        </div>
      </InstantSearch>
    </Wrapper>
  )
}

export default ProjectsArchive
```

However, I'm still getting the warning in my console, and the back button is not working :(

Edit: I also tried completely removing the stateMapping property from routing, and then the console warning doesn't appear, but unfortunately this doesn't resolve the back button issue either :(

Edit 2: The issue only appears with `next/link` Links, not with a-tags. But that defeats the purpose obviously.

**Reply by tom-radikal** [reply]
An extra question: how do the routeToState and stateToRoute relate to operations? Because I just got a mail that my quota has been used, while on Monday this was only at 16%. In the console I saw that stateToRoute triggers a lot of times, while routeToState only triggers once.

**Reply by Haroenv** [reply]
@tom-radikal, operations are the number of network requests you do. There are more calls to routing than there are to the network request, as stateToRoute is also called for example for the URLs of pagination.

For your other question, where are the NextLinks on the page? I don't see them. Do you have a reproduction on codesandbox?

**Reply by tom-radikal** [reply]
If I try it on codesandbox, the back button seems to work. The project is pretty big though, haven't taken the time to completely reproduce. I'll see if I can do that later today.

What I am wondering now though, is if it can be a problem of how I implemented the InstantSearch components. I have multiple, which are instantiated in different components. I have an instance inside my Header which works for sitewide search, which is included on all pages (via Layout component). But next to that, I have three other instances which are included on different pages (ProjectsArchive, NewsArchive, ProgrammesArchive). When I navigate to another page, that InstantSearch instance is probably unmounted. Should I maybe wrap my app or my Layout in an InstantSearch instance? And if so, how would I go about implementing multiple instances which use different indices?

Sorry for all these questions, the docs were very clear for the most part, it's just with this last part that I have struggled. I'd like to share the staging link, but since it's a site for a major Belgian broadcaster, so I don't want to share it publicly :)

**Reply by Haroenv** [reply]
I understand. Every InstantSearch instance is related to their own routing, so indeed when it unmounts, if you have multiple routings on the same page active that could cause conflicts.

Each instance has their own index as well, or not? 

If you could try to reproduce that part of the setup only (multiple InstantSearch), we can have a look and see what's the best pattern in your case.

**Reply by tom-radikal** [reply]
Each instance has their own index, except for the sitewide search in the header, that one has three (dev_project, dev_news, dev_programme). Which is implemented like this (shortened):

```
<InstantSearch searchClient={searchClient} indexName="dev_news">
  <Index
    indexName={`dev_project`}
  >...</Index>
  <Index
    indexName={`dev_news`}
  >...</Index>
  <Index
    indexName={`dev_programme`}
  >...</Index>
</InstantSearch>
```

I have to provide an `indexName` to InstantSearch, so I'm not sure if that is the correct way to implement it.

Can I use the same `instant_search` from the examples in the codesandbox for multiple InstantSearch instances? Since my current free tier is blocked because of a code error I made yesterday (page got stuck trying to set the UiState to my own data store, so I'm assuming it spammed a huge amount of requests in a short time), I can't use our own indices. The request was made to upgrade the account, but this has to go through several channels, and could take a couple of days.

**Reply by Haroenv** [reply]
yes, you can use the credentials from the demos to make an example :) 

I'd say without understanding the full code yet, that you likely want to make sure you're using only one InstantSearch instance

**Reply by tom-radikal** [reply]
For now, it seems we have to use the standard `routing={true}`, and force a page refresh, because if we use the `react-instantsearch-hooks-router-nextjs` package with singletonRouter, every request is triggered twice in the network tab. I removed every InstantSearch instance except that one, and each request (clicking a facet, going to another page, etc...) triggers twice, unless I simply use `routing={true}`.

**Reply by Haroenv** [reply]
That definitely doesn't seem correct. Again, if you could reproduce this specific situation somehow, I'd love to investigate what's exactly going on!

**Reply by tom-radikal** [reply]
I have a barebones reproduction of the issue here, removed everything but the search box:
https://codesandbox.io/s/funny-borg-9rumww?file=/components/ProjectsArchive.tsx
The pages are created at build time, so SSR and a serverUrl isn't needed for this project.

If you open the network tab, and enter anything in the search box and submit, you can see that the request is done twice.

It could very well be something that I did wrong or misunderstood from the docs, but I've been stumped for a while now :)

Thanks for checking!

**Reply by Haroenv** [reply]
Two things (algolia client, router) in your example are instantiated _inside_ the component (and thus recreated on every render), which means their cache does not get persisted. Once those two things are moved outside the component, it removes the conditional and the extra identical query:

https://codesandbox.io/s/affectionate-pine-yg1nh4?file=/components/ProjectsArchive.tsx

**Reply by tom-radikal** [reply]
Thank you so much. So obvious now that I see it 🤦 

I have done this in the other archive components as well, seems like something I should check in other projects as well. Thanks again!

**Reply by tom-radikal** [reply]
Unfortunately I discovered another issue with this, added it to this codesandbox: https://codesandbox.io/s/eloquent-shape-d6ex0c

When I do a search, then click a link, then click the back button to go back to the search page, any new searches aren't reflected in the url anymore.

**Reply by Haroenv** [reply]
I initially thought this condition would be false https://github.com/algolia/instantsearch/blob/0625c90346e32b926d6ce276a4e0b13d4fa4bf6c/packages/react-instantsearch-hooks-router-nextjs/src/index.ts#L104, but looking further shows that that's not the issue.

What seems more interesting now is that isDisposed only seems to be set to `true`, never back to `false` https://github.com/algolia/instantsearch/blob/0625c90346e32b926d6ce276a4e0b13d4fa4bf6c/packages/instantsearch.js/src/lib/routers/history.ts#L224, and thus `shouldWrite` returns false if the InstantSearch instance has "at some point" been disposed. I'm not sure how that wasn't detected before, but I'm investigating into it deeper now

**Reply by Haroenv** [reply]
latest InstantSearch version (6.41.0) has this fixed, see https://codesandbox.io/s/old-shadow-92mmm4?file=/package.json

**Reply by tom-radikal** [reply]
Thank you, that does indeed seem to solve the issue. In my own project I still have the issue that the query parameters are removed when using the back button. Not always, but sometimes. I have seen it once on the codesandbox just now, but I can't reproduce it now.

**Reply by Haroenv** [reply]
Don't hesitate to open a new issue with reproduction if you see other edge cases!

**Reply by tom-radikal** [reply]
This looks to be related with my use of an extra component to rewrite the Head (next/head) for SEO purposes.

**Reply by tom-radikal** [reply]
I was too fast, I'm still getting the error of the query params being removed. What I did notice, is that I am getting this warning every time it happens, although I have each widget that is referenced in the warning:

<img width="1265" alt="image" src="https://user-images.githubusercontent.com/94045743/223730509-40e686fe-3ea5-48a8-ac64-fd050c2fc6b7.png">

I can't seem to reproduce this on codesandbox though :(

**Reply by Haroenv** [reply]
are any of the widgets conditional? I would guess maybe they aren't mounted when InstantSearch is initially started?

**Reply by tom-radikal** [reply]
No they're all descendants of InstantSearch, none of them are conditional. I also tried playing with the `writeDelay` option, setting it to a couple of seconds for example. All the widgets are on screen before the params are removed.

**Reply by Haroenv** [reply]
could you add a debugger on the point that warning is given? if the warning is correct, it's because `widgets` is empty

**Reply by tom-radikal** [reply]
Widgets is undefined, and at that moment nothing is on screen indeed:

![image](https://user-images.githubusercontent.com/94045743/223735210-fd2badff-dc06-4fff-97fd-a65ab0de5ffd.png)

There is this, though:

<img width="400" alt="image" src="https://user-images.githubusercontent.com/94045743/223735599-f52c8d7d-8ffc-44b4-9458-4096ad7f8743.png">

**Reply by Haroenv** [reply]
I don't think we'll be able to help if you can't reproduce this in a minimal sandbox though, all I can think of is conditionals you don't see (suspense maybe?)

**Reply by tom-radikal** [reply]
Does that include outside of the InstantSearch component? I have tried, even copying a bunch of the codebase on codesandbox (https://codesandbox.io/s/staging-shadow-vw8rh6?file=/components/projects/ProjectsArchive.tsx), but I can't reproduce it there.

On my project, when I click a link with the parameters, they also get removed.

The only way I managed to get it working was to wrap my app in the InstantSearch component, but that's not possible since I have 3 Archive pages that all need to work in the same way. If there was a way to add the routing to the <Index> component instead of the <InstantSearch> one :(. 

Edit: something else that happens on my project, but not on codesandbox: when navigating away from the page, it does 7 requests to Algolia, each with a facet or query less than the previous.

**Reply by tom-radikal** [reply]
Here is one way I managed to reproduce my error. I'm getting it in all browsers with the complete project, but this might help you to pinpoint the issue:

1. Open a new instance of Firefox, with a clear cache
2. Open a new incognito window
3. Go to the page: https://codesandbox.io/s/staging-shadow-vw8rh6?file=/components/projects/ProjectsArchive.tsx
4. Click the link that's visible. It should link to the archive with the query params, and they are removed.
5. In the console you should see the warning that I was getting
